onboard/main.cpp

```diff
@@ -0,0 +1,269 @@
+  const int IN  = 0;
+  const int OUT = 1;
+  const int ERROR = 2;
+
+  // PID Variables
+  const int PIDPeriod = 10; // milliseconds
+
+  //State Machine Variables
+  enum STATES{
+    CALIBRATION,
+    RECEIVE_TRAJECTORY,
+    RUN_STATIC_CONTROL,
+    RUN_STATIC_ALL,
+    RUN_TRAJECTORY,
+    RUN_ALL_TRAJECTORIES,
+    WAIT
+  };
+  int state = STATES::WAIT;
+  int currLocalJoint = 0;
+  volatile unsigned long int timer = 0;
+
+  void stopMotors(){
+    for (int i=0; i<numOfJoints; i++){
+      joints[i].motorPWM(zeroTorque);
+      joints[i].setEnable(LOW);
+    }
+  }
+
+  //State machine functions
+  void ID_Request(){
+    commDataFromTeensy msgOut;
+    int32_t links[numOfJoints];
+    for (int i = 0; i<numOfJoints; i++){
+      links[i] = EEPROM.read(joints[i].getMemoryAddr().jointAddr);
+      msgOut.joints[i] = links[i];
+    }
+    lcm.publish(OUT, &msgOut);
+  }
+
+  void Calibration_State(){
+    if ((&joints[currLocalJoint])->CalibrationCheck() == false) {
+      error_channel error_msg;
+      error_msg.potHardwareOutOfRange = true;
+      lcm.publish(ERROR, &error_msg);
+    }
+  }
+
+  void Static_Control_State(){
+    bool OutOfRange = joints[currLocalJoint].checkOOR();
+    if (OutOfRange) {
+      error_channel error_msg;
+      error_msg.isOutOfRange = true;
+      lcm.publish(ERROR, &error_msg);
+      state = STATES::WAIT;
+    }
+    else{
+      (&joints[currLocalJoint])->PIDcontrol();
+    }
+  }
+
+//Convert from Joint number 1-12 to 0-2 for local uses
+  int convertToLocalJointNumber(int num){
+    return ((num-1)%numOfJoints);
+  }
+
+  void Static_Control_All_State(){
+    bool OutOfRange = joints[0].checkOOR()||joints[1].checkOOR()||joints[2].checkOOR();
+    if (OutOfRange) {
+      error_channel error_msg;
+      error_msg.isOutOfRange = true;
+      lcm.publish(ERROR, &error_msg);
+      state = STATES::WAIT;
+    }
+    else{
+      for (int i=0; i<numOfJoints; i++){
+        (&joints[i])->PIDcontrol();
+      }
+    }
+  }
+
+  //Timer interrupt callback, increases the timer variable and sets the flag ready for PID control
+  void timerCallback() {
+    timer++;
+    if (timer >= PIDPeriod) { //every 10 ms
+      timer = 0;
+    }
+  }
+//================================STATE_ASSIGNMENT=======================================
+  //State assignment
+void stateAssignment(int command){
+    switch (command) {
+
+      case commData2Teensy::ID_REQUEST:
+        if (state == STATES::WAIT){
+          ID_Request();
+        }
+        else{
+          error_channel msg_Out;
+          msg_Out.inappropriateCommand = true;
+          lcm.publish(OUT, &msg_Out);
+        }
+        break;
+
+      case commData2Teensy::START_CALIBRATION:
+        if (state == STATES::WAIT){
+          uint16_t val = (uint16_t)joints[currLocalJoint].readPotentiometer();
+          (&joints[currLocalJoint])->setMinPot(val);
+          (&joints[currLocalJoint])->setMaxPot(val);
+          (&joints[currLocalJoint])->setZeroPot(val);
```

Comment:
Why are you casting to a pointer and using the arrow operator? (&foo)->bar() is exactly equivlent to foo.bar()